### DISCUSSION TITLE
Timestamps for the video on kitty website

### DISCUSSION THREAD

**Discussion by waikoo** [kovidgoyal/kitty#6281] [Ideas]
Wouldn't it be great if the topics in [the video](https://sw.kovidgoyal.net/kitty/) would be listed with timestamps for easier access?

Youtube-style: if you click on the timestamp, the video jumps to that location.
This is what I came up with, feel free to modify it.

00:00 - Introduction
00:39 - Pager: View command output in same window `Ctrl+Shift+g`
01:43 - Pager: View command output in a separate window
02:14 - Pager: Utilizes shell integration in kitty
02:27 - Tab text: Display current working directory (`cwd`) and last executed command (`cmd`)
03:03 - Open files/images from ls output with mouse: `Ctrl+Shift+click` on listed file
04:04 - Open files/images from ls output with keyboard: List files with numbers and use `Ctrl+Shift+P` followed by the number
04:26 - Open files on click: Use `ls --hyperlink=auto`
05:03 - Open files on click: Configure filetypes in `open-actions.conf` in kitty
05:45 - Click on grep output to open in editor: Use `hyperlinked-grep` kitten
07:18 - SSH: Use remote file kitten to view remote files locally
08:31 - SSH: Use remote file kitten to edit remote files locally
10:01 - View images directly with `icat` kitten
10:36 - icat: Download and display images/gifs from the internet
11:03 - Live image preview in ranger with the `Kitty Graphics Protocol`
11:25 - SSH: Display images from remote server with `icat`
12:04 - Emojis in terminal with `Unicode Input Kitten`
12:54 - Window Management: Introduction
13:36 - Switch window focus: Use `Ctrl+Shift+<num>` where `<num>` is the window number
13:48 - Visually select window: Use `Ctrl+Shift+F7`
13:58 - Simultaneous input in all windows
14:15 - Interactive Kitty Shell: Use `Ctrl+Shift+Esc`
14:36 - Broadcast text with `launch --allow-remote-control kitty +kitten broadcast`
15:18 - Control kitty from other shells/programs with the `Kitty Remote Control Protocol`
15:52 - Kitty Shell: Access help with the "help" command
16:34 - Choose kitty themes interactively: Use `kitty +kitten themes -h`
17:23 - Choose kitty theme by name: Use `kitty +kitten themes [options] [theme_name]`

**Comment by kovidgoyal** [comment]
Definitely cool, but I lack the time/motivation to actually implement it, sorry. The website is generated from the files in the docs folder so a PR implementing it there is welcome. THe https://sw.kovidgoyal.net/kitty/build/ page has instructions for building the docs as well, IIRC.

**Reply by waikoo** [reply]
> Definitely cool, but I lack the time/motivation to actually implement it, sorry. The website is generated from the files in the docs folder so a PR implementing it there is welcome. THe https://sw.kovidgoyal.net/kitty/build/ page has instructions for building the docs as well, IIRC.

From the cloned project root I ran:
`python3 -m pip install -r docs/requirements.txt && make docs`
This is the error I got:
```bash
Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/boolean/.local/lib/python3.10/site-packages/sphinx/config.py", line 354, in eval_config_file
    exec(code, namespace)  # NoQA: S102
  File "/media/boolean/HDD1/GitHub/kitty/docs/conf.py", line 29, in <module>
    from kitty.conf.types import Definition, expand_opt_references  # noqa
  File "/media/boolean/HDD1/GitHub/kitty/kitty/conf/types.py", line 12, in <module>
    import kitty.conf.utils as generic_parsers
  File "/media/boolean/HDD1/GitHub/kitty/kitty/conf/utils.py", line 28, in <module>
    from ..fast_data_types import Color
ModuleNotFoundError: No module named 'kitty.fast_data_types'

make[1]: *** [Makefile:24: man] Error 2
make[1]: Leaving directory '/media/boolean/HDD1/GitHub/kitty/docs'
make: *** [Makefile:45: man] Error 2
```
How can I get hold of `kitty.fast_data_types`? Running `make html` gave me the same error from inside the `docs` folder.

**Comment by kovidgoyal** [comment]
You need to build kitty itself first with make. Only after kitty is
built can the docs be built.

On Wed, May 17, 2023 at 10:22:54AM -0700, waikoo wrote:
> > Definitely cool, but I lack the time/motivation to actually implement it, sorry. The website is generated from the files in the docs folder so a PR implementing it there is welcome. THe https://sw.kovidgoyal.net/kitty/build/ page has instructions for building the docs as well, IIRC.
> 
> From the cloned project root I ran:
> `python3 -m pip install -r docs/requirements.txt && make docs`
> This is the error I got:
> ```bash
> Configuration error:
> There is a programmable error in your configuration file:
> 
> Traceback (most recent call last):
>   File "/home/boolean/.local/lib/python3.10/site-packages/sphinx/config.py", line 354, in eval_config_file
>     exec(code, namespace)  # NoQA: S102
>   File "/media/boolean/HDD1/GitHub/kitty/docs/conf.py", line 29, in <module>
>     from kitty.conf.types import Definition, expand_opt_references  # noqa
>   File "/media/boolean/HDD1/GitHub/kitty/kitty/conf/types.py", line 12, in <module>
>     import kitty.conf.utils as generic_parsers
>   File "/media/boolean/HDD1/GitHub/kitty/kitty/conf/utils.py", line 28, in <module>
>     from ..fast_data_types import Color
> ModuleNotFoundError: No module named 'kitty.fast_data_types'
> 
> make[1]: *** [Makefile:24: man] Error 2
> make[1]: Leaving directory '/media/boolean/HDD1/GitHub/kitty/docs'
> make: *** [Makefile:45: man] Error 2
> ```
> How can I get hold of `kitty.fast_data_types`? Running `make html` gave me the same error from inside the `docs` folder.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6281#discussioncomment-5930385
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by waikoo** [comment]
@kovidgoyal Could you please advise on integrating my timestamp generating javascript module with kitty docs? 

I currently have a working prototype in a vacuum. This is how I imagined it:
A `timestamps.json` file gets looped over by a `timestamps.js` module to draw the timestamps to the screen, which can be used by the imprint of a single function call in `kitty/docs/_static/custom.js`'s already existing IIFE. Would this be okay?

However, I don't know how to couple it into the docs.

Spinning up a dev server from `kitty/docs/_build/html` doesn't draw the video element that originates from `kitty/docs/index.rst`. Can you please tell me how can I get a version of the website where I also have the output of the .rst files, so I can test things out?

![timestamps](https://i.imgur.com/V7Ri2NA.png)

Thanks!

**Comment by kovidgoyal** [comment]
On Sun, May 28, 2023 at 03:05:34PM -0700, waikoo wrote:
> @kovidgoyal Could you please advise on integrating my timestamp generating javascript module with kitty docs? 
> 
> I currently have a working prototype in a vacuum. This is how I imagined it:
> A `timestamps.json` file gets looped over by a `timestamps.js` module to draw the timestamps to the screen, which can be used by the imprint of a single function call in `kitty/docs/_static/custom.js`'s already existing IIFE. Would this be okay?

Sure, add your files to html_js_files in docs/conf.py. I am not sure
about the JSON file, it would get added as a <script> tag, you might
instead add some code to docs/conf.py in setup() that will embed the
contents of the JSON into the JS file (i.e. created a generated JS file
which contains the code from timestamp.js and the data from the JSON and
add that generated filename to html_js_files.

> 
> However, I don't know how to couple it into the docs.
> 
> Spinning up a dev server from `kitty/docs/_build/html` doesn't draw the video element that originates from `kitty/docs/index.rst`. Can you please tell me how can I get a version of the website where I also have the output of the .rst files, so I can test things out?

make develop-docs
Open http://localhost:8000 in your browser.
The video element shows up and plays, at least for me in Chrome.

**Comment by waikoo** [comment]
Thanks, I got it working! 
I made a regular .js file out of what was the JSON file.
I'd need to import this source to the module working it, so I'd need to be able to specify for the script's attribute `type` to be `module`.

Is there a way to ensure this? 
I don't know how or where could I specify it without having access to the "final" html file from the get go.
Not having two separate files would be just messy I think.

Thank you for your help!

**Comment by kovidgoyal** [comment]
See https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_js_files

**Comment by waikoo** [comment]
Opened (my first real) pull request #6321. Thank you very much for your help!
